Add semantic conventions for outgoing FaaS invocations (#862)
open-telemetry/opentelemetry-specification: specification/trace/semantic_conventions/faas.md

[Azure functions]: https://docs.microsoft.com/en-us/azure/azure-functions/manage-connections#static-clients
[Google functions]: https://cloud.google.com/functions/docs/concepts/exec#function_scope_versus_global_scope

## Incoming Invocations

This section describes incoming FaaS invocations as they are reported by the FaaS instance itself.

For incoming FaaS spans, the span kind MUST be `Server`.

| Attribute | Type | Description | Required |
|---|---|---|---|

> | `faas.coldstart` | boolean | Indicates that the serverless function is executed for the first time (aka cold start). | No |

Reviewer: Why is only this attribute "Incoming"? I assume the trigger and execution are also only applicable for incoming?

Author: Trigger applies to both (a client directly invoking the function knows what it's using for that) and execution also applies to both in case the function's endpoint/framework returns the ID in their response. I do however expect the execution being a coldstart or not to be a detail hidden by the function.

Reviewer: It's a bit inconsistent though that we use separate attribute keys (invoked vs current) for name and region but not for execution and trigger. I guess it kinda makes sense since the name/region are resource attributes that logically apply to every span already but it's a bit head scratching-inducing.
On the other hand we use http.url for both incoming and outgoing too (there is no http.client.url or something).

Author: Indeed, the `invoked*` span attributes correspond to the function's resource attributes whereas the common span attributes are the same on both spans (client and server).

## Outgoing Invocations

This section describes outgoing FaaS invocations as they are reported by a client calling a FaaS instance.

For outgoing FaaS spans, the span kind MUST be `Client`.

The values reported by the client for the attributes listed below SHOULD be equal to

> the respective [FaaS resource attributes][] and [Cloud resource attributes][],

Reviewer: ```suggestion
the corresponding [FaaS resource attributes][] and [Cloud resource attributes][],
```

Use same wording as in table header below.

> which the invoked FaaS instance reports about itself, if it's instrumented.

Reviewer: ```suggestion
which the invoked FaaS instance reports about itself, if it would be instrumented.
```

The client cannot (and should not need to) know if the remote is instrumented.

| `faas.invoked_provider` | string | The cloud provider of the invoked function. | `aws` | Yes |

> | `faas.invoked_region` | string | The cloud region of the invoked function. | `eu-central-1` | See below |

Author: I'm inviting y'all to a naming/namespacing discussion! 😀 

Please vote:

🎉 - `faas.invoked_name`, `faas.invoked_provider`, `faas_invoked_region`
🚀 - `faas.invoked.name`, `faas.invoked.provider`, `faas.invoked.region` (introducing a separate `invoked` namespace)
👀 - the above, but with `outgoing` instead of `invoked`
😕 - something else

Reviewer: I'm OK with all of them. 😃

Reviewer 2: I think invoked is fine. I would weakly prefer a namespace probably, but I have no real reason other than "feels right".

For some cloud providers, like AWS or GCP, the region in which a function is hosted is essential
to uniquely identify the function and also part of its endpoint.
Since it's part of the endpoint being called, the region is always known to clients.
In these cases, `faas.invoked_region` MUST be set accordingly.
If the region is unknown to the client or not required for identifying the invoked function,
setting `faas.invoked_region` is optional.

The values for the attributes listed above SHOULD be equal to the respective [FaaS resource attributes][]
and [Cloud resource attributes][], which the invoked FaaS instance reports about itself.

> The matching resource attributes are `faas.name`, `cloud.provider` and `cloud.region`.

Reviewer: I would try to squeeze that relation into the table.

[Azure functions]: https://docs.microsoft.com/en-us/azure/azure-functions/manage-connections#static-clients
[Google functions]: https://cloud.google.com/functions/docs/concepts/exec#function_scope_versus_global_scope

## Incoming Invocations

This section describes incoming FaaS invocations as they are reported by the FaaS instance itself.

For incoming FaaS spans, the span kind MUST be `Server`.

| Attribute | Type | Description | Required |
|---|---|---|---|
| `faas.coldstart` | boolean | Indicates that the serverless function is executed for the first time (aka cold start). | No |

## Outgoing Invocations

This section describes outgoing FaaS invocations as they are reported by a client calling a FaaS instance.

For outgoing FaaS spans, the span kind MUST be `Client`.

The values reported by the client for the attributes listed below SHOULD be equal to
the corresponding [FaaS resource attributes][] and [Cloud resource attributes][],
which the invoked FaaS instance reports about itself, if it's instrumented.

| Attribute | Corresponding resource attribute | Type | Description | Example | Required |
|---|---|---|---|---|---|

> | `faas.invoked_name` | `faas.name` | string | The name of the invoked function. | `my-function` | Yes |

Reviewer 3: Is the attribute name intentionally different here and in the resource? What's the purpose of the difference?

Author: The paragraph above the table explains that these span attributes on the client span are supposed to match the corresponding resource attributes on the function itself. Since those are different attributes (with the same values if both sides are instrumented), they have to be named differently. Please also take a look at the example at the end of the file and let me know what I could add to clarify things further.

Reviewer 3: I still do not quite understand why the name needs to be different. Do we intend that they can be populated simultaneously on the same span when the span is executed on a FaaS and represents a client call to another FaaS? If that is the case we are entering an uncharted territory which I mentioned about a while ago: the need to record 2 different Resources (the from->to scenario). If this is what we are modeling then I would want to see a generic mechanism for this (e.g. to allow representing a Resource in Span attributes). Such generic mechanism will likely be different from what we see here.

Author: One is a resource attribute and the other is a span attribute. Didn't we just recently specify names have to be unique?
It doesn't necessarily have two be two FaaS instances as I mentioned in the examples on #861. I don't know if there is a general need for modeling the remote end but for FaaS it certainly made sense to me to describe the function a client is calling. We do the same for HTTP, RPC and messaging but since FaaS instance only serve a single purpose, their properties can be modeled on the resource, which is not always the case for the others and therefore span attributes are used on the server side there as well. I see FaaS as a special case in those regards.

Reviewer 3: >One is a resource attribute and the other is a span attribute. 

Yes, they are currently. What I am saying is that we are essentially modeling an operation that may need to reference 2 locations (2 Resources). One location is the client of the function and the second location is the called function. 

To me this shows that ability to record a single Resource in a piece of telemetry data is not always sufficient. I'll stop here since I don't think we want to tackle this problem now and I don't want to derail this PR which I believe is valuable.

Author: I do see that this is conceptually done here. However, I still see FaaS as a special case where one process always serves one single purpose and thus certain information is reported on the central resource, whereas in other scenarios it is reported on the server-side spans in our current architecture.
If you think we should aim for a general solution to allow for this, I'm happy to discuss this on a separate issue and I'm glad you find this PR valuable, thanks! 🙂

Reviewer 3: >If you think we should aim for a general solution to allow for this

I think this a valid long-term goal, but I don't think we can manage to do anything reasonable by GA. Certainly no need to do anything in this PR. :-)

Function as a Service offers such flexibility that it is not possible to fully cover with semantic conventions.
When a function does not satisfy any of the aforementioned cases, a span MUST set the attribute `faas.trigger` to `"other"`.
In this case, it is responsibility of the framework or instrumentation library to define the most appropriate attributes.

## Example

This example shows the FaaS attributes for a (non-FaaS) process hosted on Google Cloud Platform (Span A with kind `Client`), which invokes a Lambda function called "my-lambda-function" in Amazon Web Services (Span B with kind `Server`).

> | Span Attribute | Resource Attribute | Span A (Client, GCP) | Span B (Server, AWS Lambda) |

Reviewer: ```suggestion
| Attribute Kind | Attribute Key | Span A (Client, GCP) | Span B (Server, AWS Lambda) |
```

With "Attribute Kind" being Resource or Span would look more readable to me.

| -------------- | ----------------------- | ---------------------- | -- |
| Resource       | `cloud.provider`        | `"gcp"`                | `"aws"` |
| Resource       | `cloud.region`          | `"europe-west3"`       | `"eu-central-1"` |
| Span           | `faas.trigger`          | `"http"`               | `"http"` |

> | Span           | `faas.invoked_name`     | `"my-lambda-function"` | n/a |

Reviewer 4: Can a client know such details? The client's just making an HTTP request, which could be lambda, or spring boot or anything.

Reviewer: I think these conventions are meant for when the client is using a specialized cloud vendor SDK like https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/lambda.html#Lambda.Client.invoke

Author: @anuraaga These are intended for cases when the client is aware that it is (directly) invoking a FaaS instance and wants to specify its details. If it is not known to the client because the function is indirectly invoked or exposed like a generic REST endpoint, they are not applicable. This is similar to a Redis database, for example, which could look like a generic REST endpoint and callers might not be aware it's directly implemented by the database - the DB semantic conventions would not be used in that case either.

@Oberon00 When using the Lambda SDK, for example, it is known that transport will be over HTTP. See https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Lambda.html, for example.
I can well imagine, however, that the transport layer might be hidden away in some libraries and their configuration might not be available to the client (or more precisely, the instrumentation), so I'll adapt the `faas.trigger` requirement to not required for clients.

Reviewer 4: Thanks for the explanation, I didn't know about that usage of lambda :) If we can describe it briefly that'd be nice, in particular the `http` trigger made me assume this is an HTTP call.

Author: Your assumption is not wrong either. Lambda functions can be invoked directly via HTTP calls using the Invoke API: https://docs.aws.amazon.com/lambda/latest/dg/API_Invoke.html

Reviewer 4: TBH I should know more about lambda than I do probably. But IIUC, you mean an HTTP call to the Invoke API, not an HTTP call directly to the lambda trigger right? I think that's different from the "http trigger", and in fact regardless of the triggering mechanism, if the SDK is used the function can be invoked. Looking at [this doc](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Lambda.html#invoke-property) I don't see any reason for the trigger to be involved when invoking the API directly, or in other words, even if the trigger is SQS, this method of invocation using the SDK technically happens over HTTP. So I think @Oberon00 has a point here, I can't find anything in the attached document to indicate we'd know what the "non-SDK trigger" is, nor does it seem like we should care when invoking the function directly, as opposed to via the trigger.

From the doc
"To invoke your function directly, use Invoke. To invoke your function in response to events in other AWS services, create an event source mapping (CreateEventSourceMapping), or configure a function trigger in the other service. For more information, see Invoking Functions."

https://docs.aws.amazon.com/lambda/latest/dg/lambda-invocation.html 

"process data automatically" refers to trigger, but not API use I think. Either you invoke manually via SDK and do know invocation name, or you invoke via trigger, where it's automatic and we can't fill this information. Does that sound right?

Author: I adapted the requirement for `faas.trigger` to no be required by clients if unknown and adapted the example accordingly. As discussed, please file an issue if you feel that the existing `faas.trigger` attribute needs to be refined (in a separate PR). Thanks!
